src/architecture/storage.md

```diff
@@ -1,262 +1,158 @@
-in a deterministic manner using [table identifiers](#table-identifiers).
+## Low-level Storage
 
-### Table Identifiers
+Exonum uses a third-party database engine to persist blockchain state
+locally. Currently the main database engine is [RocksDB][rocks-db]. It is also
+possible to plug in other engines.
 
-Every table is uniquely identified by a compound identifier, which is used
-to map table keys into a column family and its keys in the underlying
-low-level storage. A table identifier consists of 2 parts:
+To use a particular database, a minimal [`Database`][database]
+interface should be implemented for it:
 
-- **String name,** which is mapped 1-to-1 to a column family.
-  The name may contain uppercase and lowercase Latin letters, digits,
-  underscores `_`, and periods `.`. By convention, table names in services
+- Get a value by a [column family][col-family] name and a key
+- Put a new value at the specified column family / key (insert or update
+  the saved one)
+- Delete a key-value pair by a column family name / key.
+
+All the objects functionality is reduced to these atomic call types.
+
+Values of different objects and their items are stored in a single column
+family in the
+low-level storage. Keys are represented as byte sequences. Values are
+serialized according to the Protobuf serialization format. All keys are mapped
+to the low-level storage keys in a deterministic manner
+using [object identifiers](#object-identifiers).
+
+### Object Identifiers
+
+Each object in the database has its own unique identifier and metadata. The
+metadata can store various information about the internal state of the object.
+For example, for [`ListIndex`], its length is stored in the metadata.
+Metadata is represented by the [`IndexMetadata`][index-metadata] structure,
+which stores the object identifier and the metadata state.
+
+To obtain object identifiers [object pool][indexes-pool] is used. The pool
+stores identifiers of all available objects. The pool assigns identifiers to
+the objects incrementally as soon as new objects appear. The pool is
+stored in a separate column family. Key in this
+column family is the object identifier, value is the object metadata.
+
+On user level every object is uniquely identified by an address. This address
+maps the object to the identifier from the object pool mentioned before.
+An object address consists of 2 parts:
+
+- **String name,** that may contain uppercase and lowercase Latin letters,
+  digits,
+  underscores `_`, and periods `.`. By convention, object names in services
   should
   start with [the service name][service-name] and a period. For example,
-  the only table in the Cryptocurrency Tutorial is named
+  the only object in the Cryptocurrency Tutorial is named
   `cryptocurrency.wallets`,
   where `cryptocurrency` is the service name, and `wallets` is the own name
-  of the table.
+  of the object.
 - **Optional prefix** presented as a sequence of bytes (`Vec<u8>` in Rust
```

Comment:
This documentation doesn't seem to be detailed enough — above we present half a dozen of system collections that use prefixes, but don't explain what it is.